### PR TITLE
Fixing failing tests for optimized azks

### PR DESCRIPTION
**Pull request by Jasleen1**
Working through the checklist detailed here: https://github.com/novifinancial/akd/pull/189#issuecomment-1135376243

### CODE REVIEW

File: akd/src/append_only_zks.rs [hunk 1]
```diff
@@ -297,26 +266,27 @@ impl Azks {
         &self,
         storage: &S,
         label: NodeLabel,
-        epoch: u64,
+        _epoch: u64,
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:269]: nit: If we no longer need the epoch for these proofs (and it isn't going to be put back later) can we clean up the function signature?

File: akd/src/append_only_zks.rs [hunk 2]
```diff
@@ -413,39 +381,29 @@ impl Azks {
...
-                    .await?,
+                hash: to_digest::<H>(&node.hash)?,
             });
             return Ok((unchanged, leaves));
         }
-        if node.get_birth_epoch() > end_epoch {
-            // really you shouldn't even be here. Later do error checking
-            return Ok((unchanged, leaves));
-        }
+        // if node.get_birth_epoch() > end_epoch {
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:388]: nit: From the comment it looks like this wasn't necessary anyways. Can we cleanup this artifact?

File: akd/src/append_only_zks.rs [hunk 3]
```diff
@@ -725,6 +680,37 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    #[allow(dead_code)]
```
Comment (slawlor) [inline akd/src/append_only_zks.rs:684]: nit: Do we need the dead code attribute? is it not in a `#[cfg(test)]` block? If not can we move them to a

```
#[cfg(test)]
mod tests {

...
}
```

File: akd/src/history_tree_node.rs [hunk 4]
```diff
@@ -141,50 +157,24 @@ impl HistoryTreeNode {
     pub(crate) async fn get_from_storage<S: Storage + Send + Sync>(
         storage: &S,
         key: &NodeKey,
-        current_epoch: u64,
+        _current_epoch: u64,
```
Comment (slawlor) [inline akd/src/history_tree_node.rs:160]: nit: Remove the unused (unless we're going to add it back later)?

File: akd/src/history_tree_node.rs [hunk 5]
```diff
@@ -141,50 +157,24 @@ impl HistoryTreeNode {
...
     pub(crate) async fn batch_get_from_storage<S: Storage + Send + Sync>(
         storage: &S,
         keys: &[NodeKey],
-        current_epoch: u64,
+        _current_epoch: u64,
     ) -> Result<Vec<HistoryTreeNode>, StorageError> {
         let node_records: Vec<DbRecord> = storage.batch_get::<HistoryTreeNode>(keys).await?;
         let mut nodes = Vec::<HistoryTreeNode>::new();
-        for (i, node) in node_records.into_iter().enumerate() {
+        for (_i, node) in node_records.into_iter().enumerate() {
```
Comment (slawlor) [inline akd/src/history_tree_node.rs:175]: nit: Don't enumerate if we don't need the iterator index. so just ```for node in node_records.into_iter()```

File: akd/src/history_tree_node.rs [hunk 6]
```diff
@@ -266,373 +250,209 @@ impl HistoryTreeNode {
...
-                    self.update_hash::<_, H>(storage, epoch).await?;
-                    self.write_to_storage(storage).await?;
-                } else {
-                    debug!("BEGIN retrieve self");
-                    *self = HistoryTreeNode::get_from_storage(storage, &NodeKey(self.label), epoch)
-                        .await?;
-                }
-                debug!("END insert single leaf (dir_self = None)");
-                Ok(())
+                let result = match child_node {
```
Comment (slawlor) [inline akd/src/history_tree_node.rs:316]: nit: Remove the assignment of result, and just have the match statement end the fn.

File: akd/src/history_tree_node.rs [hunk 7]
```diff
@@ -266,373 +250,209 @@ impl HistoryTreeNode {
...
-            },
-        }
+        Ok(())
     }
 
     /// Inserts a child into this node, adding the state to the state at this epoch,
     /// without updating its own hash.
-    #[async_recursion]
-    pub(crate) async fn set_child<S: Storage + Sync + Send, H: Hasher>(
+    // #[async_recursion]
```
Comment (slawlor) [inline akd/src/history_tree_node.rs:399]: nit: just delete this

File: akd/src/history_tree_node.rs [hunk 8]
```diff
@@ -645,270 +465,142 @@ impl HistoryTreeNode {
...
+                if dir == 0 {
+                    Ok(self.left_child)
+                } else if dir == 1 {
+                    Ok(self.right_child)
                 } else {
-                    // Indeterminate, we are somewhere above the
-                    // birth epoch but we're less than the "last" epoch.
-                    // db query is necessary
-                    None
+                    panic!(
```
Comment (slawlor) [inline akd/src/history_tree_node.rs:509]: nit: can we return `AkdError` instead of panicing?

File: akd/src/storage/types.rs [hunk 9]
```diff
@@ -20,8 +19,8 @@ pub enum StorageType {
     Azks = 1,
     /// HistoryTreeNode
     HistoryTreeNode = 2,
-    /// HistoryNodeState
-    HistoryNodeState = 3,
+    /// EOZ: HistoryNodeState = 3 was removed from here.
```
Comment (slawlor) [inline akd/src/storage/types.rs:22]: To answer the question in the comment: We're going to do a DB break anyways, we can update this to 3. If we were adding types and trying for a db upgrade, then yeah. But until we release "version 1" we don't have much to worry about if we're going to break all backwards compat.

**Review by slawlor — COMMENTED** [review]
So generally, it seems like we've either deleted a lot of tests or at least disabled them. Is this considered WIP? Like eventually we'll restore them right?

Otherwise it's just a lot of tiny styling nits and whatnot, but this looks really, really good! Thanks!

**Comment by eozturk1** [comment]
@slawlor I'll re-iterate over the deleted/commented out tests and bring the ones we will need back. Tracked in #208.

File: akd/src/append_only_zks.rs [hunk 10]
```diff
@@ -790,6 +787,66 @@ mod tests {
...
+    async fn test_nonmembership_proof_v_small() -> Result<(), AkdError> {
```
Comment (eozturk1) [inline akd/src/append_only_zks.rs:792]: nit: very small?

File: akd/src/append_only_zks.rs [hunk 11]
```diff
@@ -790,6 +787,66 @@ mod tests {
...
+        for i in 0..num_nodes {
+            let mut label_arr = [0u8; 32];
+            label_arr[31] = u8::from(i);
+            let label = NodeLabel::new(label_arr, 256u32);
+            let mut input = [0u8; 32];
+            input[31] = u8::from(i);
+            let hash = Blake3Digest::new(input);
+            let node = Node::<Blake3> { label, hash };
+            insertion_set.push(node);
+        }
```
Comment (eozturk1) [inline akd/src/append_only_zks.rs:806]: Could we move this part to a separate helper function?

File: akd/src/append_only_zks.rs [hunk 12]
```diff
@@ -790,6 +787,66 @@ mod tests {
...
+    #[tokio::test]
+    #[allow(dead_code)]
+    async fn test_nonmembership_proof_v_small() -> Result<(), AkdError> {
+        let num_nodes = 2;
+
+        let mut insertion_set: Vec<Node<Blake3>> = vec![];
+
+        for i in 0..num_nodes {
+            let mut label_arr = [0u8; 32];
+            label_arr[31] = u8::from(i);
```
Comment (eozturk1) [inline akd/src/append_only_zks.rs:799]: Why do we do this?

File: akd/src/append_only_zks.rs (hunk 10 above)
Comment (eozturk1) [inline akd/src/append_only_zks.rs:792]: Can you remove `#[allow(dead_code)]` and add a description for the test?

File: akd/src/append_only_zks.rs (hunk 11 above)
Comment (eozturk1) [reply akd/src/append_only_zks.rs:806]: What about `generate_random_node()`? 

If we want a specific property for the node, let's add this as a comment.

File: akd/src/append_only_zks.rs [hunk 13]
```diff
@@ -790,6 +787,66 @@ mod tests {
...
+        let proof = azks.get_non_membership_proof(&db, search_label, 2).await?;
+
+        verify_nonmembership::<Blake3>(azks.get_root_hash::<_, Blake3>(&db).await?, &proof)?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[allow(dead_code)]
+    async fn test_nonmembership_proof_small() -> Result<(), AkdError> {
```
Comment (eozturk1) [inline akd/src/append_only_zks.rs:821]: Same as above function.

File: akd/src/directory.rs [hunk 14]
```diff
@@ -263,17 +258,19 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
...
         let commitment_key = self.derive_commitment_key::<H>().await?;
         let plaintext_value = lookup_info.value_state.plaintext_val;
-
+        let existence_vrf = self
+            .vrf
+            .get_label_proof::<H>(&uname, false, current_version)
+            .await?;
+        let commitment_label = self
+            .vrf
+            .get_node_label_from_vrf_pf::<H>(existence_vrf)
```
Comment (eozturk1) [inline akd/src/directory.rs:267]: nit: We can keep this as `proof`, no need to shorten :)

File: akd/src/tests.rs [hunk 15]
```diff
@@ -57,7 +57,7 @@ async fn test_simple_publish() -> Result<(), AkdError> {
     Ok(())
 }
 
-// #[tokio::test]
+#[tokio::test]
 #[allow(dead_code)]
```
Comment (eozturk1) [inline akd/src/tests.rs:61]: We should remove `#[allow(dead_code)]`s since we have enabled the tests.

File: akd/src/utils.rs [hunk 16]
```diff
@@ -94,11 +95,10 @@ pub(crate) fn get_commitment_proof<H: Hasher>(
...
 pub(crate) fn commit_value<H: Hasher>(
     commitment_key: &[u8],
-    label: &AkdLabel,
-    version: u64,
+    label: &NodeLabel,
     value: &AkdValue,
 ) -> H::Digest {
-    let proof = get_commitment_proof::<H>(commitment_key, label, version, value);
+    let proof = get_commitment_proof::<H>(commitment_key, label, value);
     H::hash(&[i2osp_array(value), i2osp_array(&proof.as_bytes())].concat())
```
Comment (eozturk1) [inline akd/src/utils.rs:102]: Should we use `i2osp_array` before each hash operation?

File: akd/src/append_only_zks.rs [hunk 17]
```diff
@@ -790,6 +787,66 @@ mod tests {
...
+            let mut input = [0u8; 32];
+            input[31] = u8::from(i);
+            let hash = Blake3Digest::new(input);
+            let node = Node::<Blake3> { label, hash };
+            insertion_set.push(node);
+        }
+        let db = AsyncInMemoryDatabase::new();
+        let mut azks = Azks::new::<_, Blake3>(&db).await?;
+        let search_label = insertion_set[0].label;
+        azks.batch_insert_leaves::<_, Blake3>(&db, insertion_set.clone()[1..2].to_vec())
```
Comment (eozturk1) [inline akd/src/append_only_zks.rs:810]: Let's separate the node we are searching for from the nodes we are inserting into the tree.

```
let node_to_search = generate_random_node();
let insertion_set = generate_nodes(); // uses generate_random_node()
azks.batch_insert_leaves(insertion_set);
let proof = azks.get_non_membership_proof(node_to_search.label);
// Verify proof
```

**Comment by Jasleen1** [comment]
This PR addresses #144 and #207, except the last task in the checklist of #207.

File: akd/src/proof_structs.rs [hunk 18]
```diff
@@ -92,23 +92,40 @@ impl<H: Hasher> Clone for NonMembershipProof<H> {
...
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
 #[cfg_attr(feature = "serde", serde(bound = ""))]
 pub struct AppendOnlyProof<H: Hasher> {
+    /// Proof for a single epoch being append-only
+    pub proofs: Vec<SingleAppendOnlyProof<H>>,
+    /// Epochs over which this audit is being performed
+    pub epochs: Vec<u64>,
+}
+
+// FIXME: need to fix the documentation
```
Comment (kevinlewi) [inline akd/src/proof_structs.rs:111]: Does this still need to be addressed?

File: akd_client/src/tests.rs [hunk 19]
```diff
@@ -392,6 +407,7 @@ async fn test_history_proof_multiple_epochs() -> Result<(), AkdError> {
 }
 
 #[tokio::test]
+#[allow(dead_code)]
```
Comment (kevinlewi) [inline akd_client/src/tests.rs:410]: Can we remove this?

File: akd_client/src/tests.rs [hunk 20]
```diff
@@ -436,6 +449,7 @@ async fn test_history_proof_single_epoch() -> Result<(), AkdError> {
...
```
Comment (kevinlewi) [inline akd_client/src/tests.rs:452]: Can we remove this?

**Review by kevinlewi — COMMENTED** [review]
There are a couple of allow(dead_code)'s. Can they be removed?

File: akd_client/src/tests.rs (hunk 19 above)
Comment (eozturk1) [reply akd_client/src/tests.rs:410]: All `#[allow(dead_code)]`s should be removed after enabling the tests. I mentioned this in a few other places. @Jasleen1 said she would do a clean up before merging.

**Comment by kevinlewi** [comment]
Ok cool. Also, we want to merge this into the `eozturk1/optimized-azks` branch, not `main`, right?

File: akd/src/proof_structs.rs (hunk 18 above)
Comment (Jasleen1) [reply akd/src/proof_structs.rs:111]: Will be addressing in a cleanup PR

**Pull request closed without merging**